### ISSUE TITLE
Unable to catch type error when using spread operator

### ISSUE THREAD

**Issue by richardliu-db** [microsoft/TypeScript#56927]
### 🔎 Search Terms

spread operator

https://github.com/microsoft/TypeScript/issues/38657#issuecomment-1644445738

### 🕗 Version & Regression Information

- typescript version: 5.2.2

### 💻 Code

```ts
type Type = {
  a: string
}

const foo = ({ a }: Type) => {
  console.log(a)
}

foo({ a: 'hello', b: 'world' }) // Compile error. Expected

foo({ ...{ a: 'hello', b: 'world' } }) // No compile error. Not expected.
```

### 🙁 Actual behavior

The spread operator can fool the type check, as shown in the line with "No compile error. Not expected.".

### 🙂 Expected behavior

The type check should show compile error as `b` field is not expected

**Comment by fatcerberus** [comment]
Duplicate of https://github.com/microsoft/TypeScript/issues/39998.

Note that an excess property error is not a type error *per se* - objects can legally have more properties than the type declares.  This is more of a lint check and you shouldn’t rely on it to enforce runtime invariants.  See #12936.
